From 195934b76a4bf8b8b22ee6dc9d95cb98f55d8d9d Mon Sep 17 00:00:00 2001
From: Gerrit Pape <pape@smarden.org>
Date: Sun, 26 Aug 2007 09:37:32 +0000
Subject: [PATCH] util-vserver: vsched fails to read config directory

Jan H Eringa wrote:
> Use of /etc/vservers/<vserver name>/sched/.... causes vsched to blow up
> |my-server# vsched --xid 40000 --dir /etc/vservers/test-1/sched
> |vsched: readdir(): Bad address
>
>
> Which I suspect is the cause of....
> |my-server# /etc/init.d/util-vserver start
> |Fixing visibility of /proc entries for Linux-VServer guests...readdir():
> Bad
>  address
> |readdir(): Bad address
[...]

This is due to dietlibc not properly aligning the buffer it's handing to
sys_readdir. Check your dmesg output, and it should say something like
Kernel unaligned access at TPC[4b7268] filldir64+0x70/0x134

http://people.linux-vserver.org/~dhozac/p/m/delta-dietdirent-fix01.diff
fixes it, but you'll have to rebuild util-vserver as well after applying
that patch to dietlibc.

--
Daniel Hokka Zakrisson
---
 dietdirent.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/dietdirent.h b/dietdirent.h
index dbd7206..487d5a9 100644
--- a/dietdirent.h
+++ b/dietdirent.h
@@ -1,8 +1,8 @@
 #include <sys/shm.h>
 
 struct __dirstream {
-  int fd;
   char buf[PAGE_SIZE-(sizeof (int)*3)];
+  int fd;
   unsigned int num;
   unsigned int cur;
 };				/* stream data from opendir() */
-- 
1.5.3.GIT

